SDWebImage/SDImageCache.m

```diff
@@ -86,7 +86,7 @@ - (nonnull instancetype)initWithNamespace:(nonnull NSString *)ns
         
         NSAssert([config.diskCacheClass conformsToProtocol:@protocol(SDDiskCache)], @"Custom disk cache class must conform to `SDDiskCache` protocol");
         _diskCache = [[config.diskCacheClass alloc] initWithCachePath:_diskCachePath config:_config];
-        [self moveOldDefaultDiskPathOfFilesToNewDefaultDiskPath];
+        [self migrateCacheDiskDirectory];
```

Comment:
...emmm. Am I miss something ?

I means, move all code, into `[SDDiskCache initWithCachePath:config:]` method. Don't need to modify anything in `SDImageCache.m`

During init method of `SDDiskCache`, use dispatch_once + global queue to only move once.

Reply:
@dreampiggy emm, I know you opinion, so I put the reasons in [here](https://github.com/rs/SDWebImage/pull/2412#issuecomment-409621832). I can explain my thoughts further.
Why I put the default cache path in `SDImageCache` is we generate the default cache path in `SDImageCache`, right? Let's see how we generate in `SDImageCache`:

```
// class of SDImageCache
- (instancetype)init {
    return [self initWithNamespace:@"default"];
}
- (nonnull instancetype)initWithNamespace:(nonnull NSString *)ns {
    NSString *path = [self makeDiskCachePath:ns]; // first component disk cache path 
    return [self initWithNamespace:ns diskCacheDirectory:path];
}
- (nonnull instancetype)initWithNamespace:(nonnull NSString *)ns
                       diskCacheDirectory:(nonnull NSString *)directory
                                   config:(nullable SDImageCacheConfig *)config {
    if ((self = [super init])) {
        NSString *namespacePrefix = config.namespacePrefix;
        if (!namespacePrefix) {
            namespacePrefix = @"";
        }
        NSString *fullNamespace = [namespacePrefix stringByAppendingString:ns]; 
      
        // Init the disk cache
        if (directory != nil) {
            _diskCachePath = [directory stringByAppendingPathComponent:fullNamespace];
        } else {
            NSString *path = [self makeDiskCachePath:ns];
            _diskCachePath = path;
        }
    return self;
}
```

We combine many components to generate default cache disk path. The `SDDiskCache` is just a class that do disk operations, like access, delete, move, right? other words, it doesn't need to know what's the disk path it is, its only job is do disk operation, I think compare old and new disk path is not what it should do. If you still think we should put all these things to `SDDiskCache`, should we need to hard code the default path? or extract the methods of default cache path related methods into a single class to let both `SDImageCache` and `SDDiskCache` access it?